ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -1617,16 +1623,42 @@ void handleResponse(ResponseInfo responseInfo, PutResponse putResponse) {
       checkAndMaybeComplete();
     }
 
+    /**
+     * Process response if it was rejected due to quota compliance.
+     * @param correlationId correlation id of the request.
+     * @param replicaId {@link ReplicaId} of the request.
+     */
+    private void processQuotaRejectedResponse(int correlationId, ReplicaId replicaId) {
+      logger.debug("{}: PutRequest with response correlationId {} was rejected because quota was exceeded.",
+          loggingContext, correlationId);
+      setChunkException(new RouterException("QuotaExceeded", RouterErrorCode.TooManyRequests));
+      onErrorResponse(replicaId, TrackedRequestFinalState.QUOTA_REJECTED, false);
+      checkAndMaybeComplete();
```

Comment:
I don't think you are check is the chunk is rejected because of quota in checkAndMaybeComplete().

Reply:
Since `onErrorResponse` sets the operation exception, we actually don't need to check fail on quota anywhere. I have updated other places to reflect that. Thanks for pointing this out.